File: frontend/types/pricing.ts
export type PropertyInput = {
  propertyName: string;
  address: string;
  city: string;
  state: string;
  zipCode: string;
  propertyType: string;
  bedrooms: number;
  bathrooms: number;
  areaSqft: number;
  yearBuilt: number;
  furnishingStatus: string;
  floorNumber: number;
  parkingSpaces: number;
  parkingAvailability: string;
  hasGym: boolean;
  hasPool: boolean;
  petFriendly: boolean;
  gatedCommunity: boolean;
  nearbyMetroDistance: number;
  amenities: string[];
};

export type NeighborhoodMetricsInput = {
  schoolQualityScore: number;
  floodRiskScore: number;
  transitScore: number;
  roadNoiseScore: number;
  neighborhoodAppealScore: number;
};

export type RecommendationPayload = {
  property: PropertyInput;
  neighborhood: NeighborhoodMetricsInput;
  analystNotes?: string;
};

export type ComparableProperty = {
  id: number;
  propertyName: string;
  address: string;
  monthlyRent: number;
  distanceMiles: number;
  similarityScore: number;
  matchPercentage: number;
  selectionReason: string;
  bedrooms: number;
  bathrooms: number;
  areaSqft: number;
  keyMatchingFactors: string[];
  weight: number;
  status: string;
};

export type RecommendationResponse = {
  recommendationId: number;
  recommendedRent: number;
  rentLow: number;
  rentHigh: number;
  confidenceScore: number;
  reasoningSummary: string;
  pricePerSqft: number;
  marketTrendIndicator: string;
  pricingFactors: { name: string; value: number }[];
  neighborhoodSummary: { label: string; score: number }[];
  comparables: ComparableProperty[];
};

export type ComparableActionType =
  | "remove"
  | "increase_weight"
  | "decrease_weight"
  | "mark_irrelevant";

export type UpdateRecommendationPayload = {
  recommendationId: number;
  feedbackText?: string;
  comparableActions?: {
    comparableId: string;
    action: ComparableActionType;
  }[];
};

export type FeedbackResponse = {
  updatedRecommendation: RecommendationResponse;
  parsedIntent: string;
  pricingAdjustment: number;
};

File: backend/app/schemas/pricing.py
from uuid import UUID

from pydantic import BaseModel

from app.schemas.property import NeighborhoodPayload, PropertyPayload


class RecommendationRequest(BaseModel):
    property: PropertyPayload
    neighborhood: NeighborhoodPayload
    analyst_notes: str | None = None


class ComparableResponse(BaseModel):
    id: int
    property_name: str
    address: str
    monthly_rent: float
    bedrooms: int
    bathrooms: float
    area_sqft: int
    distance_miles: float
    similarity_score: float
    match_percentage: float
    selection_reason: str
    key_matching_factors: list[str]
    weight: float = 1.0
    status: str = "active"


class RecommendationResponse(BaseModel):
    recommendation_id: int
    recommended_rent: float
    rent_low: float
    rent_high: float
    confidence_score: float
    price_per_sqft: float
    reasoning_summary: str
    market_trend_indicator: str
    pricing_factors: list[dict]
    neighborhood_summary: list[dict]
    comparables: list[ComparableResponse]


class FeedbackRequest(BaseModel):
    recommendation_id: int
    feedback_text: str


class ComparableAction(BaseModel):
    comparable_id: UUID | int
    action: str


class RecommendationUpdateRequest(BaseModel):
    recommendation_id: int
    feedback_text: str | None = None
    comparable_actions: list[ComparableAction] = []


class FeedbackResponse(BaseModel):
    updated_recommendation: RecommendationResponse
    parsed_intent: str
    pricing_adjustment: float
